src/function/aggregate/nested/list.cpp

```diff
@@ -5,49 +5,665 @@
+
+	ListSegment *segment = nullptr;
+
+	// determine segment
+	if (!linked_list->last_segment) {
+		// empty linked list, create the first (and last) segment
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreateSegment(allocator, owning_vector, capacity, input);
+		linked_list->first_segment = segment;
+		linked_list->last_segment = segment;
+
+	} else if (linked_list->last_segment->capacity == linked_list->last_segment->count) {
+		// the last segment of the linked list is full, create a new one and append it
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreateSegment(allocator, owning_vector, capacity, input);
+		linked_list->last_segment->next = segment;
+		linked_list->last_segment = segment;
+
+	} else {
+		// the last segment of the linked list is not full, append the data to it
+		segment = linked_list->last_segment;
+	}
+
+	D_ASSERT(segment);
+	return segment;
+}
+
+ListSegment *ListFun::GetCharSegment(Allocator &allocator, vector<unique_ptr<AllocatedData>> &owning_vector,
+                                     LinkedList *linked_list) {
+
+	ListSegment *segment = nullptr;
+
+	// determine segment
+	if (!linked_list->last_segment) {
+		// empty linked list, create the first (and last) char segment
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreatePrimitiveSegment(allocator, owning_vector, capacity, LogicalTypeId::VARCHAR);
+		linked_list->first_segment = segment;
+		linked_list->last_segment = segment;
+
+	} else if (linked_list->last_segment->capacity == linked_list->last_segment->count) {
+		// the last segment of the linked list is full, create a new char segment and append it
+		auto capacity = GetCapacityForNewSegment(linked_list);
+		segment = CreatePrimitiveSegment(allocator, owning_vector, capacity, LogicalTypeId::VARCHAR);
+		linked_list->last_segment->next = segment;
+		linked_list->last_segment = segment;
+
+	} else {
+		// the last segment of the linked list is not full, append the data to it
+		segment = linked_list->last_segment;
+	}
+
+	D_ASSERT(segment);
+	return segment;
+}
+
+void ListFun::WriteDataToSegment(Allocator &allocator, vector<unique_ptr<AllocatedData>> &owning_vector,
+                                 ListSegment *segment, Vector &input, idx_t &entry_idx, idx_t &count) {
+
+	// get the vector data and the source index of the entry that we want to write
+	UnifiedVectorFormat input_data;
+	input.ToUnifiedFormat(count, input_data);
+	auto source_idx = input_data.sel->get_index(entry_idx);
+
+	// write null validity
+	auto null_mask = GetNullMask(segment);
+	auto is_null = !input_data.validity.RowIsValid(source_idx);
+	null_mask[segment->count] = is_null;
+
+	// write value
+	if (input.GetType().InternalType() == PhysicalType::VARCHAR) {
+
+		// set the length of this string
+		auto str_length_data = GetListLengthData(segment);
+		uint64_t str_length = 0;
+
+		// get the string
+		string str = "";
```

Comment:
`std::string` is slow, is this required? Can we not use/read the `string_t` directly?